<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Guorui Zhong">
<meta name="description" content="Comat models batch effects as multiplicative and additive noise to the biological signal and uses a Bayesian framework to fit linear models that factor such noise out of the readouts.">
<title>Guorui Zhong - Combat to remove batch effects in high-throughput experiments</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../">
<script src="../../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><style>

      .quarto-title-block .quarto-title-banner {
        background: #212529;
      }
</style>
</head>
<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../../../index.html">
    <span class="navbar-title">Guorui Zhong</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../../../index.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../blog/index.html" rel="" target="">
 <span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../crispr/index.html" rel="" target="">
 <span class="menu-text">CRISPR</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../keyboard/index.html" rel="" target="">
 <span class="menu-text">Keyboard</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../python/index.html" rel="" target="">
 <span class="menu-text">Python</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../r/index.html" rel="" target="">
 <span class="menu-text">R</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://scholar.google.com/citations?user=g0MfTPMAAAAJ&amp;hl=en" rel="" target=""><i class="bi bi-journal-check" role="img" aria-label="journal-check">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/guorui-zhong-754761185/" rel="" target=""><i class="bi bi-linkedin" role="img" aria-label="linkedin">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/guoruizhong" rel="" target=""><i class="bi bi-github" role="img" aria-label="github">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Combat to remove batch effects in high-throughput experiments</h1>
                  <div>
        <div class="description">
          Comat models batch effects as multiplicative and additive noise to the biological signal and uses a Bayesian framework to fit linear models that factor such noise out of the readouts.
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">batch effect</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p><a href="https://guoruizhong.github.io/">Guorui Zhong</a> <a href="https://orcid.org/0000-0002-1955-2847" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">Wednesday, August 9, 2023</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">Thursday, November 9, 2023</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Contents</h2>
   
  <ul>
<li><a href="#what-is-batch-effect" id="toc-what-is-batch-effect" class="nav-link active" data-scroll-target="#what-is-batch-effect">What is batch effect?</a></li>
  <li><a href="#why-is-batch-effect-a-problem" id="toc-why-is-batch-effect-a-problem" class="nav-link" data-scroll-target="#why-is-batch-effect-a-problem">Why is batch effect a problem</a></li>
  <li><a href="#sva-package-for-removing-batch-effects" id="toc-sva-package-for-removing-batch-effects" class="nav-link" data-scroll-target="#sva-package-for-removing-batch-effects">SVA package for removing batch effects</a></li>
  <li><a href="#setting-up-the-data-from-an-expressionset" id="toc-setting-up-the-data-from-an-expressionset" class="nav-link" data-scroll-target="#setting-up-the-data-from-an-expressionset">Setting up the data from an ExpressionSet</a></li>
  <li><a href="#eestimate-batch-effect-and-other-artifacts-using-sva" id="toc-eestimate-batch-effect-and-other-artifacts-using-sva" class="nav-link" data-scroll-target="#eestimate-batch-effect-and-other-artifacts-using-sva">Eestimate batch effect and other artifacts using sva</a></li>
  <li><a href="#adjusting-for-surrogate-variables-using-f.pvalue-function" id="toc-adjusting-for-surrogate-variables-using-f.pvalue-function" class="nav-link" data-scroll-target="#adjusting-for-surrogate-variables-using-f.pvalue-function">Adjusting for surrogate variables using f.pvalue function</a></li>
  <li><a href="#adjusting-for-surrogate-variables-using-the-limma" id="toc-adjusting-for-surrogate-variables-using-the-limma" class="nav-link" data-scroll-target="#adjusting-for-surrogate-variables-using-the-limma">Adjusting for surrogate variables using the limma</a></li>
  <li><a href="#adjusting-for-known-batches-using-combat" id="toc-adjusting-for-known-batches-using-combat" class="nav-link" data-scroll-target="#adjusting-for-known-batches-using-combat">Adjusting for known batches using Combat</a></li>
  <li><a href="#removing-known-batch-effects-with-a-linear-model" id="toc-removing-known-batch-effects-with-a-linear-model" class="nav-link" data-scroll-target="#removing-known-batch-effects-with-a-linear-model">Removing known batch effects with a linear model</a></li>
  <li><a href="#comparing-combat-and-linear-adjustment" id="toc-comparing-combat-and-linear-adjustment" class="nav-link" data-scroll-target="#comparing-combat-and-linear-adjustment">Comparing ComBat and linear adjustment</a></li>
  <li><a href="#combat-seq-for-batch-adjustment-on-rna-seq-count-data" id="toc-combat-seq-for-batch-adjustment-on-rna-seq-count-data" class="nav-link" data-scroll-target="#combat-seq-for-batch-adjustment-on-rna-seq-count-data">Combat-seq for batch adjustment on RNA-seq count data</a></li>
  <li><a href="#variance-filtering-to-speed-computations" id="toc-variance-filtering-to-speed-computations" class="nav-link" data-scroll-target="#variance-filtering-to-speed-computations">Variance filtering to speed computations</a></li>
  <li><a href="#applying-the-fsva-function-to-remove-batch-effects-for-prediction" id="toc-applying-the-fsva-function-to-remove-batch-effects-for-prediction" class="nav-link" data-scroll-target="#applying-the-fsva-function-to-remove-batch-effects-for-prediction">Applying the fsva function to remove batch effects for prediction</a></li>
  <li><a href="#reference" id="toc-reference" class="nav-link" data-scroll-target="#reference">Reference</a></li>
  </ul></nav>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><section id="what-is-batch-effect" class="level2"><h2 class="anchored" data-anchor-id="what-is-batch-effect">What is batch effect?</h2>
<p>Batch effects are sub-groubs of measurements that have qualitatively different behaviour across conditions and unrelated to the biological or scientific variables in a study. - Experiments run on different time - Different technician - Differnt doing the sample preparation - Sequencing at a differnt institute - Prepare the sample at different timepoint - Different lots of reagents, chips or instruments …</p>
<p>All of these factors are considered to be factors as batches</p>
</section><section id="why-is-batch-effect-a-problem" class="level2"><h2 class="anchored" data-anchor-id="why-is-batch-effect-a-problem">Why is batch effect a problem</h2>
<ul>
<li>Not biological difference</li>
<li>Just the other factors are taking actions, and are showing differences</li>
<li>Most of the time, batch effect deviate our finding from the real biological relevant information</li>
</ul>
<p>Try our best to eliminate batch effect to get a real difference between conditions!</p>
</section><section id="sva-package-for-removing-batch-effects" class="level2"><h2 class="anchored" data-anchor-id="sva-package-for-removing-batch-effects">SVA package for removing batch effects</h2>
<p>The assumption is that you will be trying to analyze the association between the variables of interest and gene expression, adjusting for the adjustment variables</p>
</section><section id="setting-up-the-data-from-an-expressionset" class="level2"><h2 class="anchored" data-anchor-id="setting-up-the-data-from-an-expressionset">Setting up the data from an ExpressionSet</h2>
<ul>
<li>The data should be a matrix with features in the rows and samples in the columns.</li>
<li>The data should be standardized before applying for correction.</li>
</ul>
<p><img src="sample.png" class="img-fluid"></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/">here</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">sva</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">bladderbatch</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pamr</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/">limma</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/therneau/survival">survival</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">bladderdata</span><span class="op">)</span></span>
<span><span class="co">### variable data</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html">pData</a></span><span class="op">(</span><span class="va">bladderEset</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></span>
<span><span class="co">## [1] "data.frame"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pheno</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##              sample outcome batch cancer</span></span>
<span><span class="co">## GSM71019.CEL      1  Normal     3 Normal</span></span>
<span><span class="co">## GSM71020.CEL      2  Normal     2 Normal</span></span>
<span><span class="co">## GSM71021.CEL      3  Normal     2 Normal</span></span>
<span><span class="co">## GSM71022.CEL      4  Normal     3 Normal</span></span>
<span><span class="co">## GSM71023.CEL      5  Normal     3 Normal</span></span>
<span><span class="co">### expression data matrix</span></span>
<span><span class="va">edata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html">exprs</a></span><span class="op">(</span><span class="va">bladderEset</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">edata</span><span class="op">)</span></span>
<span><span class="co">## [1] "matrix" "array"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">edata</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##           GSM71019.CEL GSM71020.CEL GSM71021.CEL GSM71022.CEL GSM71023.CEL</span></span>
<span><span class="co">## 1007_s_at    10.115170     8.628044     8.779235     9.248569    10.256841</span></span>
<span><span class="co">## 1053_at       5.345168     5.063598     5.113116     5.179410     5.181383</span></span>
<span><span class="co">## 117_at        6.348024     6.663625     6.465892     6.116422     5.980457</span></span>
<span><span class="co">## 121_at        8.901739     9.439977     9.540738     9.254368     8.798086</span></span>
<span><span class="co">## 1255_g_at     3.967672     4.466027     4.144885     4.189338     4.078509</span></span>
<span><span class="co">## 1294_at       7.775183     7.110154     7.248430     7.017220     7.896419</span></span>
<span><span class="co">### create full model matrix, including both the adjustment variables and the variable of interest (cancer status)</span></span>
<span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">cancer</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">pheno</span><span class="op">)</span></span>
<span><span class="co">### Create null model matrix, ontains only the adjustment variables.</span></span>
<span><span class="va">mod0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">1</span>, data <span class="op">=</span> <span class="va">pheno</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="eestimate-batch-effect-and-other-artifacts-using-sva" class="level2"><h2 class="anchored" data-anchor-id="eestimate-batch-effect-and-other-artifacts-using-sva">Eestimate batch effect and other artifacts using sva</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### identify the number of latent factors that need to be estimated</span></span>
<span><span class="va">n_sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/num.sv.html">num.sv</a></span><span class="op">(</span><span class="va">edata</span>, <span class="va">mod</span>, method <span class="op">=</span> <span class="st">"leek"</span><span class="op">)</span></span>
<span><span class="va">n_sv</span></span>
<span><span class="co">## [1] 2</span></span>
<span><span class="co">## estimate the surrogate variables</span></span>
<span><span class="va">svaobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/sva.html">sva</a></span><span class="op">(</span><span class="va">edata</span>, <span class="va">mod</span>, <span class="va">mod0</span>, n.sv <span class="op">=</span> <span class="va">n_sv</span><span class="op">)</span></span>
<span><span class="co">## Number of significant surrogate variables is:  2 </span></span>
<span><span class="co">## Iteration (out of 5 ):1  2  3  4  5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">svaobj</span><span class="op">)</span></span>
<span><span class="co">## List of 4</span></span>
<span><span class="co">##  $ sv       : num [1:57, 1:2] -0.02717 -0.00618 0.07737 -0.00114 -0.02062 ...</span></span>
<span><span class="co">##  $ pprob.gam: num [1:22283] 0.999 0.977 1 1 1 ...</span></span>
<span><span class="co">##  $ pprob.b  : num [1:22283] 0.9956 0.9998 0.0702 0.9998 1 ...</span></span>
<span><span class="co">##  $ n.sv     : num 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">svaobj</span><span class="op">$</span><span class="va">sv</span> <span class="op">~</span> <span class="va">pheno</span><span class="op">$</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Response Y1 :</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## lm(formula = Y1 ~ pheno$batch)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residuals:</span></span>
<span><span class="co">##      Min       1Q   Median       3Q      Max </span></span>
<span><span class="co">## -0.26953 -0.11076  0.00787  0.10399  0.19069 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##              Estimate Std. Error t value Pr(&gt;|t|)</span></span>
<span><span class="co">## (Intercept) -0.018470   0.038694  -0.477    0.635</span></span>
<span><span class="co">## pheno$batch  0.006051   0.011253   0.538    0.593</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residual standard error: 0.1345 on 55 degrees of freedom</span></span>
<span><span class="co">## Multiple R-squared:  0.00523,    Adjusted R-squared:  -0.01286 </span></span>
<span><span class="co">## F-statistic: 0.2891 on 1 and 55 DF,  p-value: 0.5929</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Response Y2 :</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## lm(formula = Y2 ~ pheno$batch)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residuals:</span></span>
<span><span class="co">##      Min       1Q   Median       3Q      Max </span></span>
<span><span class="co">## -0.23973 -0.07467 -0.02157  0.08116  0.25629 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##              Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">## (Intercept)  0.121112   0.034157   3.546 0.000808 ***</span></span>
<span><span class="co">## pheno$batch -0.039675   0.009933  -3.994 0.000194 ***</span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residual standard error: 0.1187 on 55 degrees of freedom</span></span>
<span><span class="co">## Multiple R-squared:  0.2248, Adjusted R-squared:  0.2107 </span></span>
<span><span class="co">## F-statistic: 15.95 on 1 and 55 DF,  p-value: 0.0001945</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">svaobj</span><span class="op">$</span><span class="va">sv</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">~</span> <span class="va">pheno</span><span class="op">$</span><span class="va">batch</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="va">svaobj</span><span class="op">$</span><span class="va">sv</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/jitter.html">jitter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">batch</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## null device </span></span>
<span><span class="co">##           1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="adjusting-for-surrogate-variables-using-f.pvalue-function" class="level2"><h2 class="anchored" data-anchor-id="adjusting-for-surrogate-variables-using-f.pvalue-function">Adjusting for surrogate variables using f.pvalue function</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### calculate the parametric F-test p-values for each row of a data matrix</span></span>
<span><span class="va">pvalues</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/f.pvalue.html">f.pvalue</a></span><span class="op">(</span><span class="va">edata</span>, <span class="va">mod</span>, <span class="va">mod0</span><span class="op">)</span></span>
<span><span class="co">### adjust them for multiple teseting</span></span>
<span><span class="va">qvalues</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">pvalues</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Nearly 70% of the genes are strongly differentially expressed at an FDR of less than 5% between groups.<br>
This number seems artifically high.</p>
<p>Get the adjusted p-values and q-values accounting for surrogate variables.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### include the surrogate variables</span></span>
<span><span class="va">modsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">mod</span>, <span class="va">svaobj</span><span class="op">$</span><span class="va">sv</span><span class="op">)</span></span>
<span><span class="va">mod0sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">mod0</span>, <span class="va">svaobj</span><span class="op">$</span><span class="va">sv</span><span class="op">)</span></span>
<span><span class="va">pvalues_sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/f.pvalue.html">f.pvalue</a></span><span class="op">(</span><span class="va">edata</span>, <span class="va">modsv</span>, <span class="va">mod0sv</span><span class="op">)</span></span>
<span><span class="va">qvalues_sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">pvalues</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="adjusting-for-surrogate-variables-using-the-limma" class="level2"><h2 class="anchored" data-anchor-id="adjusting-for-surrogate-variables-using-the-limma">Adjusting for surrogate variables using the limma</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### fit the linear model with the surrogate variabels</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span><span class="va">edata</span>, <span class="va">modsv</span><span class="op">)</span></span>
<span><span class="co">### compute the contrasts between cancer/normal terms</span></span>
<span><span class="va">contrast_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span></span>
<span>  <span class="st">"C1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">svaobj</span><span class="op">$</span><span class="va">n.sv</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="st">"C2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">svaobj</span><span class="op">$</span><span class="va">n.sv</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">fit_contrasts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html">contrasts.fit</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">contrast_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### calculate the test statistics</span></span>
<span><span class="va">eb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">fit_contrasts</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTableF</a></span><span class="op">(</span><span class="va">eb</span>, adjust <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="co">## topTableF is obsolete and will be removed in a future version of limma. Please considering using topTable instead.</span></span>
<span><span class="co">##                    C1          C2   AveExpr        F      P.Value    adj.P.Val</span></span>
<span><span class="co">## 207783_x_at -13.45607  0.26592268 12.938786 8622.529 1.207531e-69 1.419929e-65</span></span>
<span><span class="co">## 201492_s_at -13.27594  0.15357702 13.336090 8605.649 1.274450e-69 1.419929e-65</span></span>
<span><span class="co">## 208834_x_at -12.76411  0.06134018 13.160201 6939.501 4.749368e-67 3.527673e-63</span></span>
<span><span class="co">## 212869_x_at -13.77957  0.26008165 13.452076 6593.346 1.939773e-66 1.080599e-62</span></span>
<span><span class="co">## 212284_x_at -13.59977  0.29135767 13.070844 5495.716 2.893287e-64 1.289423e-60</span></span>
<span><span class="co">## 208825_x_at -12.70979  0.08250821 13.108072 5414.741 4.350100e-64 1.615555e-60</span></span>
<span><span class="co">## 211445_x_at -10.15890 -0.06633356  9.853817 5256.114 9.845076e-64 3.133969e-60</span></span>
<span><span class="co">## 213084_x_at -12.59345  0.03015520 13.046529 4790.107 1.260201e-62 3.510132e-59</span></span>
<span><span class="co">## 201429_s_at -13.33686  0.28358293 12.941208 4464.995 8.675221e-62 2.147888e-58</span></span>
<span><span class="co">## 214327_x_at -12.60146  0.20934783 11.832607 4312.087 2.257025e-61 5.029329e-58</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="adjusting-for-known-batches-using-combat" class="level2"><h2 class="anchored" data-anchor-id="adjusting-for-known-batches-using-combat">Adjusting for known batches using Combat</h2>
<p><code>ComBat</code> returns a “cleaned” data matrix after batch effects have been removed. Here we pass a model matrix with any known adjustment variables and a second parameter that is the batch variable.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### get known batch variable</span></span>
<span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="va">pheno</span><span class="op">$</span><span class="va">batch</span></span>
<span><span class="co">### model matrix</span></span>
<span><span class="va">modcombat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">1</span>, data <span class="op">=</span> <span class="va">pheno</span><span class="op">)</span></span>
<span><span class="va">modcancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">cancer</span>, data <span class="op">=</span> <span class="va">pheno</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Using parametric empirical Bayesian adjustments.</span></span>
<span><span class="va">combat_edata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/ComBat.html">ComBat</a></span><span class="op">(</span></span>
<span>    dat <span class="op">=</span> <span class="va">edata</span>,</span>
<span>    batch <span class="op">=</span> <span class="va">batch</span>, </span>
<span>    mod <span class="op">=</span> <span class="va">modcombat</span>,</span>
<span>    par.prior <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># performs parametric empirical Bayesian adjustment</span></span>
<span>    <span class="co"># par.prior = FALSE, # performs non-parametric empirical Bayesian adjustment</span></span>
<span>    prior.plots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    mean.only <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># only adjust the mean of the batch effects across batches</span></span>
<span>    ref.batch <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">### significance analysis</span></span>
<span><span class="va">pvalue_combat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/f.pvalue.html">f.pvalue</a></span><span class="op">(</span><span class="va">combat_edata</span>, <span class="va">mod</span>, <span class="va">mod0</span><span class="op">)</span></span>
<span><span class="va">qvalue_combat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">pvalue_combat</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">combat_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lmfit.html">lm.fit</a></span><span class="op">(</span><span class="va">modcancer</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">combat_edata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">combat_fit</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span>, col <span class="op">=</span> <span class="fl">4</span>, breaks <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="removing-known-batch-effects-with-a-linear-model" class="level2"><h2 class="anchored" data-anchor-id="removing-known-batch-effects-with-a-linear-model">Removing known batch effects with a linear model</h2>
<p>In this case, we use two models. - One with the variable we care about - cancer status - the other is just the known adjustment variables, in this case, we assume none</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mod_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">cancer</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">batch</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">pheno</span><span class="op">)</span></span>
<span><span class="va">mod0_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">batch</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">pheno</span><span class="op">)</span></span>
<span><span class="va">pvalues_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/f.pvalue.html">f.pvalue</a></span><span class="op">(</span><span class="va">edata</span>, <span class="va">mod_batch</span>, <span class="va">mod0_batch</span><span class="op">)</span></span>
<span><span class="va">qvalues_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">pvalues_batch</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lmfit.html">lm.fit</a></span><span class="op">(</span><span class="va">mod_batch</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">edata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span>, col <span class="op">=</span> <span class="fl">4</span>, breaks <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="comparing-combat-and-linear-adjustment" class="level2"><h2 class="anchored" data-anchor-id="comparing-combat-and-linear-adjustment">Comparing ComBat and linear adjustment</h2>
<p>We can compare the estimated coefficients from Combat and linear adjustment by looking at the right coefficients for each model.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">fit</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span>, <span class="va">combat_fit</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span>, col <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"Linear Model"</span>, ylab <span class="op">=</span> <span class="st">"Combat"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, col <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## null device </span></span>
<span><span class="co">##           1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Add the surrogate variables to the model matrix and perform model fit</span></span>
<span><span class="va">modsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">mod</span>, <span class="va">svaobj</span><span class="op">$</span><span class="va">sv</span><span class="op">)</span></span>
<span><span class="va">fitsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lmfit.html">lm.fit</a></span><span class="op">(</span><span class="va">modsv</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">edata</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Compare the fit from surrogate variable analysis to the other two</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">fitsv</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>, <span class="va">combat_fit</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span>, </span>
<span>  col <span class="op">=</span> <span class="fl">2</span>, xlab <span class="op">=</span> <span class="st">"SVA"</span>, ylab <span class="op">=</span> <span class="st">"Combat"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, col <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">fitsv</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>, <span class="va">fit</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span>, </span>
<span>  col <span class="op">=</span> <span class="fl">2</span>, xlab <span class="op">=</span> <span class="st">"SVA"</span>, ylab <span class="op">=</span> <span class="st">"Linear model"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, col <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## null device </span></span>
<span><span class="co">##           1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="combat-seq-for-batch-adjustment-on-rna-seq-count-data" class="level2"><h2 class="anchored" data-anchor-id="combat-seq-for-batch-adjustment-on-rna-seq-count-data">Combat-seq for batch adjustment on RNA-seq count data</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### expression matrix</span></span>
<span><span class="va">count_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html">rnbinom</a></span><span class="op">(</span><span class="fl">400</span>, size <span class="op">=</span> <span class="fl">10</span>, prob <span class="op">=</span>  <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">50</span>, ncol <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span>
<span><span class="co">### batch</span></span>
<span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### adjust</span></span>
<span><span class="va">adjusted</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/ComBat_seq.html">ComBat_seq</a></span><span class="op">(</span></span>
<span>  <span class="va">count_matrix</span>, </span>
<span>  batch <span class="op">=</span> <span class="va">batch</span>, </span>
<span>  group <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Found 2 batches</span></span>
<span><span class="co">## Using null model in ComBat-seq.</span></span>
<span><span class="co">## Adjusting for 0 covariate(s) or covariate level(s)</span></span>
<span><span class="co">## Estimating dispersions</span></span>
<span><span class="co">## Fitting the GLM model</span></span>
<span><span class="co">## Shrinkage off - using GLM estimates for parameters</span></span>
<span><span class="co">## Adjusting the data</span></span>
<span></span>
<span><span class="co">### specify one biological variable</span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">adjusted_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/ComBat_seq.html">ComBat_seq</a></span><span class="op">(</span></span>
<span>  <span class="va">count_matrix</span>, </span>
<span>  batch <span class="op">=</span> <span class="va">batch</span>, </span>
<span>  group <span class="op">=</span> <span class="va">group</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Found 2 batches</span></span>
<span><span class="co">## Using full model in ComBat-seq.</span></span>
<span><span class="co">## Adjusting for 1 covariate(s) or covariate level(s)</span></span>
<span><span class="co">## Estimating dispersions</span></span>
<span><span class="co">## Fitting the GLM model</span></span>
<span><span class="co">## Shrinkage off - using GLM estimates for parameters</span></span>
<span><span class="co">## Adjusting the data</span></span>
<span></span>
<span><span class="co">### multiple biological variabels</span></span>
<span><span class="va">cov1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">cov2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">covar_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">cov1</span>, <span class="va">cov2</span><span class="op">)</span></span>
<span><span class="va">adjusted_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/ComBat_seq.html">ComBat_seq</a></span><span class="op">(</span></span>
<span>  <span class="va">count_matrix</span>, </span>
<span>  batch <span class="op">=</span> <span class="va">batch</span>, </span>
<span>  group <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  covar_mod <span class="op">=</span> <span class="va">covar_mat</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Found 2 batches</span></span>
<span><span class="co">## Using null model in ComBat-seq.</span></span>
<span><span class="co">## Adjusting for 2 covariate(s) or covariate level(s)</span></span>
<span><span class="co">## Estimating dispersions</span></span>
<span><span class="co">## Fitting the GLM model</span></span>
<span><span class="co">## Shrinkage off - using GLM estimates for parameters</span></span>
<span><span class="co">## Adjusting the data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="variance-filtering-to-speed-computations" class="level2"><h2 class="anchored" data-anchor-id="variance-filtering-to-speed-computations">Variance filtering to speed computations</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">n_sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/num.sv.html">num.sv</a></span><span class="op">(</span><span class="va">edata</span>, <span class="va">mod</span>, vfilter <span class="op">=</span> <span class="fl">2000</span>, method <span class="op">=</span> <span class="st">"leek"</span><span class="op">)</span></span>
<span><span class="va">svaobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/sva.html">sva</a></span><span class="op">(</span><span class="va">edata</span>, <span class="va">mod</span>, <span class="va">mod0</span>, n.sv <span class="op">=</span> <span class="va">n_sv</span>, vfilter <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="co">## Number of significant surrogate variables is:  46 </span></span>
<span><span class="co">## Iteration (out of 5 ):1  2  3  4  5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="applying-the-fsva-function-to-remove-batch-effects-for-prediction" class="level2"><h2 class="anchored" data-anchor-id="applying-the-fsva-function-to-remove-batch-effects-for-prediction">Applying the fsva function to remove batch effects for prediction</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### seprate data to traning and test set</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">12354</span><span class="op">)</span></span>
<span><span class="va">train_indicator</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">57</span>, size <span class="op">=</span> <span class="fl">30</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">test_indicator</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">57</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="va">train_indicator</span><span class="op">]</span></span>
<span><span class="va">train_data</span> <span class="op">&lt;-</span> <span class="va">edata</span><span class="op">[</span>, <span class="va">train_indicator</span><span class="op">]</span></span>
<span><span class="va">test_data</span> <span class="op">&lt;-</span> <span class="va">edata</span><span class="op">[</span>, <span class="va">test_indicator</span><span class="op">]</span></span>
<span><span class="va">train_pheno</span> <span class="op">&lt;-</span> <span class="va">pheno</span><span class="op">[</span><span class="va">train_indicator</span>, <span class="op">]</span></span>
<span><span class="va">test_pheno</span> <span class="op">&lt;-</span> <span class="va">pheno</span><span class="op">[</span><span class="va">test_indicator</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">### pamr package can be used to train a predictive model and test that prediction</span></span>
<span><span class="va">mydata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">train_data</span>, y <span class="op">=</span> <span class="va">train_pheno</span><span class="op">$</span><span class="va">cancer</span><span class="op">)</span></span>
<span><span class="va">mytrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pamr/man/pamr.train.html">pamr.train</a></span><span class="op">(</span><span class="va">mydata</span><span class="op">)</span></span>
<span><span class="co">## 123456789101112131415161718192021222324252627282930</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/pamr/man/pamr.predict.html">pamr.predict</a></span><span class="op">(</span><span class="va">mytrain</span>, <span class="va">test_data</span>, threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>  <span class="va">test_pheno</span><span class="op">$</span><span class="va">cancer</span></span>
<span><span class="op">)</span></span>
<span><span class="co">##         </span></span>
<span><span class="co">##          Biopsy Cancer Normal</span></span>
<span><span class="co">##   Biopsy      4      0      0</span></span>
<span><span class="co">##   Cancer      0     15      0</span></span>
<span><span class="co">##   Normal      1      4      3</span></span>
<span></span>
<span><span class="co">### calculate surrogate variables for training set</span></span>
<span><span class="va">train_mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">cancer</span>, data <span class="op">=</span> <span class="va">train_pheno</span><span class="op">)</span></span>
<span><span class="va">train_mod0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">1</span>, data <span class="op">=</span> <span class="va">train_pheno</span><span class="op">)</span></span>
<span><span class="va">train_sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/sva.html">sva</a></span><span class="op">(</span><span class="va">train_data</span>, <span class="va">train_mod</span>, <span class="va">train_mod0</span><span class="op">)</span></span>
<span><span class="co">## Number of significant surrogate variables is:  5 </span></span>
<span><span class="co">## Iteration (out of 5 ):1  2  3  4  5</span></span>
<span></span>
<span><span class="co">### adjust both the training data and the test data</span></span>
<span><span class="va">fsvaobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/sva/man/fsva.html">fsva</a></span><span class="op">(</span><span class="va">train_data</span>, <span class="va">train_mod</span>, <span class="va">train_sv</span>, <span class="va">test_data</span><span class="op">)</span></span>
<span><span class="va">mydata_sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">fsvaobj</span><span class="op">$</span><span class="va">db</span>, y <span class="op">=</span> <span class="va">train_pheno</span><span class="op">$</span><span class="va">cancer</span><span class="op">)</span></span>
<span><span class="va">mytrain_sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pamr/man/pamr.train.html">pamr.train</a></span><span class="op">(</span><span class="va">mydata_sv</span><span class="op">)</span></span>
<span><span class="co">## 123456789101112131415161718192021222324252627282930</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="reference" class="level2"><h2 class="anchored" data-anchor-id="reference">Reference</h2>
<ul>
<li><a href="http://jtleek.com/genstats/inst/doc/02_13_batch-effects.html">Batch effects and confounders</a></li>
<li><a href="https://www.bioconductor.org/packages/release/bioc/vignettes/sva/inst/doc/sva.pdf">The SVA package for removing batch effects and other unwanted variation in high-throughput experiments</a></li>
<li><a href="https://evayiwenwang.github.io/Managing_batch_effects/">Managing batch effects</a></li>
</ul>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center"><span class="faux-block"><i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> 2021-2024 Guorui Zhong ~ Made with <a href="https://www.r-project.org/"><i class="fa-brands fa-r-project" aria-label="r-project"></i></a> and <a href="https://quarto.org/">Quarto</a></span></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>