<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Guorui Zhong">
<meta name="description" content="The package that must to learn for scRNA-seq data analysis.">
<title>Learning Seruat for scRNA-seq data analysis â€“ Guorui Zhong</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../../">
<script src="../../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-html/quarto-syntax-highlighting-cee8642dd6abb73bb51e0193ec9c0780.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../../site_libs/bootstrap/bootstrap-a122eddc3d35102178d4fe042f5b75d5.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><style>

      .quarto-title-block .quarto-title-banner {
        background: #212529;
      }
</style>
</head>
<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../../index.html">
    <span class="navbar-title">Guorui Zhong</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../../../index.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../blog/index.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../crispr/index.html"> 
<span class="menu-text">CRISPR</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../keyboard/index.html"> 
<span class="menu-text">Keyboard</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../python/index.html"> 
<span class="menu-text">Python</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../../r/index.html"> 
<span class="menu-text">R</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://scholar.google.com/citations?user=g0MfTPMAAAAJ&amp;hl=en"> <i class="bi bi-journal-check" role="img" aria-label="journal-check">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/guorui-zhong-754761185/"> <i class="bi bi-linkedin" role="img" aria-label="linkedin">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/guoruizhong"> <i class="bi bi-github" role="img" aria-label="github">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full"><div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Learning Seruat for scRNA-seq data analysis</h1>
                  <div>
        <div class="description">
          The package that must to learn for scRNA-seq data analysis.
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">r</div>
                <div class="quarto-category">seurat</div>
                <div class="quarto-category">scRNA-seq</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p><a href="https://guoruizhong.github.io/">Guorui Zhong</a> <a href="https://orcid.org/0000-0002-1955-2847" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">Saturday, November 25, 2023</p>
      </div>
    </div>
    
      <div>
      <div class="quarto-title-meta-heading">Modified</div>
      <div class="quarto-title-meta-contents">
        <p class="date-modified">Saturday, December 16, 2023</p>
      </div>
    </div>
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Contents</h2>
   
  <ul>
<li><a href="#load-packages-and-data" id="toc-load-packages-and-data" class="nav-link active" data-scroll-target="#load-packages-and-data">Load packages and data</a></li>
  <li>
<a href="#preprocess-data" id="toc-preprocess-data" class="nav-link" data-scroll-target="#preprocess-data">Preprocess data</a>
  <ul>
<li><a href="#qc" id="toc-qc" class="nav-link" data-scroll-target="#qc">QC</a></li>
  <li><a href="#normalizing-the-data" id="toc-normalizing-the-data" class="nav-link" data-scroll-target="#normalizing-the-data">Normalizing the data</a></li>
  <li><a href="#highly-variable-features" id="toc-highly-variable-features" class="nav-link" data-scroll-target="#highly-variable-features">Highly variable features</a></li>
  <li><a href="#scaling-the-data" id="toc-scaling-the-data" class="nav-link" data-scroll-target="#scaling-the-data">Scaling the data</a></li>
  </ul>
</li>
  <li>
<a href="#perform-linear-dimensional-reduction" id="toc-perform-linear-dimensional-reduction" class="nav-link" data-scroll-target="#perform-linear-dimensional-reduction">Perform linear dimensional reduction</a>
  <ul>
<li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#visualize-it" id="toc-visualize-it" class="nav-link" data-scroll-target="#visualize-it">Visualize it</a></li>
  </ul>
</li>
  <li>
<a href="#determine-the-dimensionality-of-the-dataset" id="toc-determine-the-dimensionality-of-the-dataset" class="nav-link" data-scroll-target="#determine-the-dimensionality-of-the-dataset">Determine the dimensionality of the dataset</a>
  <ul>
<li><a href="#jackstrawplot" id="toc-jackstrawplot" class="nav-link" data-scroll-target="#jackstrawplot">JackStrawPlot</a></li>
  <li><a href="#elbow-plot" id="toc-elbow-plot" class="nav-link" data-scroll-target="#elbow-plot">Elbow plot</a></li>
  <li><a href="#significant-related-genes" id="toc-significant-related-genes" class="nav-link" data-scroll-target="#significant-related-genes">Significant related genes</a></li>
  </ul>
</li>
  <li><a href="#cluster-the-cells" id="toc-cluster-the-cells" class="nav-link" data-scroll-target="#cluster-the-cells">Cluster the cells</a></li>
  <li>
<a href="#run-non-linear-dimensional-reduction" id="toc-run-non-linear-dimensional-reduction" class="nav-link" data-scroll-target="#run-non-linear-dimensional-reduction">Run non-linear dimensional reduction</a>
  <ul>
<li><a href="#umap" id="toc-umap" class="nav-link" data-scroll-target="#umap">UMAP</a></li>
  <li><a href="#tsne" id="toc-tsne" class="nav-link" data-scroll-target="#tsne">tSNE</a></li>
  <li><a href="#compare" id="toc-compare" class="nav-link" data-scroll-target="#compare">Compare</a></li>
  </ul>
</li>
  <li>
<a href="#finding-cluster-biomarkers" id="toc-finding-cluster-biomarkers" class="nav-link" data-scroll-target="#finding-cluster-biomarkers">Finding cluster biomarkers</a>
  <ul>
<li><a href="#find-clusters" id="toc-find-clusters" class="nav-link" data-scroll-target="#find-clusters">Find clusters</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization">Visualization</a></li>
  </ul>
</li>
  <li><a href="#assign-cell-type-identity-to-clusters" id="toc-assign-cell-type-identity-to-clusters" class="nav-link" data-scroll-target="#assign-cell-type-identity-to-clusters">Assign cell type identity to clusters</a></li>
  </ul></nav>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content"><section id="load-packages-and-data" class="level2"><h2 class="anchored" data-anchor-id="load-packages-and-data">Load packages and data</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Library packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/">here</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR">SingleR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/">ggrepel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load PBMC dataset</span></span>
<span><span class="va">pbmc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="st">"./learn/2023_scRNA_Seurat/pbmc3k/hg19/"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Initialize the seurat boject witht raw (non-normalized data)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">pbmc_data</span>, project <span class="op">=</span> <span class="st">"pbmc3k"</span>, min.cells <span class="op">=</span> <span class="fl">3</span>, min.features <span class="op">=</span> <span class="fl">200</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># View the data</span></span>
<span><span class="va">pbmc</span></span>
<span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 13714 features across 2700 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (13714 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">pbmc_data</span><span class="op">)</span></span>
<span><span class="co">## [1] 32738  2700</span></span>
<span></span>
<span><span class="co"># Example a few genes in the first thirty cells</span></span>
<span><span class="va">pbmc_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TCL1A"</span>, <span class="st">"MS4A1"</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span></span>
<span><span class="co">## 3 x 30 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##                                                                    </span></span>
<span><span class="co">## CD3D  4 . 10 . . 1 2 3 1 . . 2 7 1 . . 1 3 . 2  3 . . . . . 3 4 1 5</span></span>
<span><span class="co">## TCL1A . .  . . . . . . 1 . . . . . . . . . . .  . 1 . . . . . . . .</span></span>
<span><span class="co">## MS4A1 . 6  . . . . . . 1 1 1 . . . . . . . . . 36 1 2 . . 2 . . . .</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="preprocess-data" class="level2"><h2 class="anchored" data-anchor-id="preprocess-data">Preprocess data</h2>
<section id="qc" class="level3"><h3 class="anchored" data-anchor-id="qc">QC</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># The [[ operator can add columns to object metadata. This is a great place to stash QC stats</span></span>
<span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Show QC metrics for the first 5 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##                  orig.ident nCount_RNA nFeature_RNA percent.mt</span></span>
<span><span class="co">## AAACATACAACCAC-1     pbmc3k       2419          779  3.0177759</span></span>
<span><span class="co">## AAACATTGAGCTAC-1     pbmc3k       4903         1352  3.7935958</span></span>
<span><span class="co">## AAACATTGATCAGC-1     pbmc3k       3147         1129  0.8897363</span></span>
<span><span class="co">## AAACCGTGCTTCCG-1     pbmc3k       2639          960  1.7430845</span></span>
<span><span class="co">## AAACCGTGTATGCG-1     pbmc3k        980          521  1.2244898</span></span>
<span></span>
<span><span class="co"># Visualize QC metrics as a violin plot</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co"># FeatureScatter is typically used to visualize feature-feature relationships, but can be used</span></span>
<span><span class="co"># for anything calculated by the object, i.e. columns in object metadata, PC scores etc.</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"nFeature_RNA"</span><span class="op">)</span></span>
<span><span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-2-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">pbmc</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="normalizing-the-data" class="level3"><h3 class="anchored" data-anchor-id="normalizing-the-data">Normalizing the data</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>, scale.factor <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span><span class="co"># pbmc &lt;- NormalizeData(pbmc)</span></span>
<span><span class="co"># pbmc@assays$RNA@counts is the raw count data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co">## Formal class 'Seurat' [package "SeuratObject"] with 13 slots</span></span>
<span><span class="co">##   ..@ assays      :List of 1</span></span>
<span><span class="co">##   .. ..$ RNA:Formal class 'Assay5' [package "SeuratObject"] with 8 slots</span></span>
<span><span class="co">##   .. .. .. ..@ layers    :List of 2</span></span>
<span><span class="co">##   .. .. .. .. ..$ counts:Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ i       : int [1:2238732] 29 73 80 148 163 184 186 227 229 230 ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ p       : int [1:2639] 0 779 2131 3260 4220 4741 5522 6304 7094 7626 ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ Dim     : int [1:2] 13714 2638</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">##   .. .. .. .. .. .. .. ..$ : NULL</span></span>
<span><span class="co">##   .. .. .. .. .. .. .. ..$ : NULL</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ x       : num [1:2238732] 1 1 2 1 1 1 1 41 1 1 ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ factors : list()</span></span>
<span><span class="co">##   .. .. .. .. ..$ data  :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ i       : int [1:2238732] 29 73 80 148 163 184 186 227 229 230 ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ p       : int [1:2639] 0 779 2131 3260 4220 4741 5522 6304 7094 7626 ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ Dim     : int [1:2] 13714 2638</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">##   .. .. .. .. .. .. .. ..$ : NULL</span></span>
<span><span class="co">##   .. .. .. .. .. .. .. ..$ : NULL</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ x       : num [1:2238732] 1.64 1.64 2.23 1.64 1.64 ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..@ factors : list()</span></span>
<span><span class="co">##   .. .. .. ..@ cells     :Formal class 'LogMap' [package "SeuratObject"] with 1 slot</span></span>
<span><span class="co">##   .. .. .. .. .. ..@ .Data: logi [1:2638, 1:2] TRUE TRUE TRUE TRUE TRUE TRUE ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##   .. .. .. .. .. .. .. ..$ : chr [1:2638] "AAACATACAACCAC-1" "AAACATTGAGCTAC-1" "AAACATTGATCAGC-1" "AAACCGTGCTTCCG-1" ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. .. ..$ : chr [1:2] "counts" "data"</span></span>
<span><span class="co">##   .. .. .. .. .. ..$ dim     : int [1:2] 2638 2</span></span>
<span><span class="co">##   .. .. .. .. .. ..$ dimnames:List of 2</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..$ : chr [1:2638] "AAACATACAACCAC-1" "AAACATTGAGCTAC-1" "AAACATTGATCAGC-1" "AAACCGTGCTTCCG-1" ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..$ : chr [1:2] "counts" "data"</span></span>
<span><span class="co">##   .. .. .. ..@ features  :Formal class 'LogMap' [package "SeuratObject"] with 1 slot</span></span>
<span><span class="co">##   .. .. .. .. .. ..@ .Data: logi [1:13714, 1:2] TRUE TRUE TRUE TRUE TRUE TRUE ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##   .. .. .. .. .. .. .. ..$ : chr [1:13714] "AL627309.1" "AP006222.2" "RP11-206L10.2" "RP11-206L10.9" ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. .. ..$ : chr [1:2] "counts" "data"</span></span>
<span><span class="co">##   .. .. .. .. .. ..$ dim     : int [1:2] 13714 2</span></span>
<span><span class="co">##   .. .. .. .. .. ..$ dimnames:List of 2</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..$ : chr [1:13714] "AL627309.1" "AP006222.2" "RP11-206L10.2" "RP11-206L10.9" ...</span></span>
<span><span class="co">##   .. .. .. .. .. .. ..$ : chr [1:2] "counts" "data"</span></span>
<span><span class="co">##   .. .. .. ..@ default   : int 1</span></span>
<span><span class="co">##   .. .. .. ..@ assay.orig: chr(0) </span></span>
<span><span class="co">##   .. .. .. ..@ meta.data :'data.frame':  13714 obs. of  0 variables</span></span>
<span><span class="co">##   .. .. .. ..@ misc      : Named list()</span></span>
<span><span class="co">##   .. .. .. ..@ key       : chr "rna_"</span></span>
<span><span class="co">##   ..@ meta.data   :'data.frame': 2638 obs. of  4 variables:</span></span>
<span><span class="co">##   .. ..$ orig.ident  : Factor w/ 1 level "pbmc3k": 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##   .. ..$ nCount_RNA  : num [1:2638] 2419 4903 3147 2639 980 ...</span></span>
<span><span class="co">##   .. ..$ nFeature_RNA: int [1:2638] 779 1352 1129 960 521 781 782 790 532 550 ...</span></span>
<span><span class="co">##   .. ..$ percent.mt  : num [1:2638] 3.02 3.79 0.89 1.74 1.22 ...</span></span>
<span><span class="co">##   ..@ active.assay: chr "RNA"</span></span>
<span><span class="co">##   ..@ active.ident: Factor w/ 1 level "pbmc3k": 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##   .. ..- attr(*, "names")= chr [1:2638] "AAACATACAACCAC-1" "AAACATTGAGCTAC-1" "AAACATTGATCAGC-1" "AAACCGTGCTTCCG-1" ...</span></span>
<span><span class="co">##   ..@ graphs      : list()</span></span>
<span><span class="co">##   ..@ neighbors   : list()</span></span>
<span><span class="co">##   ..@ reductions  : list()</span></span>
<span><span class="co">##   ..@ images      : list()</span></span>
<span><span class="co">##   ..@ project.name: chr "pbmc3k"</span></span>
<span><span class="co">##   ..@ misc        : list()</span></span>
<span><span class="co">##   ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1</span></span>
<span><span class="co">##   .. ..$ : int [1:3] 5 0 1</span></span>
<span><span class="co">##   ..@ commands    :List of 1</span></span>
<span><span class="co">##   .. ..$ NormalizeData.RNA:Formal class 'SeuratCommand' [package "SeuratObject"] with 5 slots</span></span>
<span><span class="co">##   .. .. .. ..@ name       : chr "NormalizeData.RNA"</span></span>
<span><span class="co">##   .. .. .. ..@ time.stamp : POSIXct[1:1], format: "2023-12-16 20:46:32"</span></span>
<span><span class="co">##   .. .. .. ..@ assay.used : chr "RNA"</span></span>
<span><span class="co">##   .. .. .. ..@ call.string: chr [1:2] "NormalizeData(pbmc, normalization.method = \"LogNormalize\", " "    scale.factor = 10000)"</span></span>
<span><span class="co">##   .. .. .. ..@ params     :List of 5</span></span>
<span><span class="co">##   .. .. .. .. ..$ assay               : chr "RNA"</span></span>
<span><span class="co">##   .. .. .. .. ..$ normalization.method: chr "LogNormalize"</span></span>
<span><span class="co">##   .. .. .. .. ..$ scale.factor        : num 10000</span></span>
<span><span class="co">##   .. .. .. .. ..$ margin              : num 1</span></span>
<span><span class="co">##   .. .. .. .. ..$ verbose             : logi TRUE</span></span>
<span><span class="co">##   ..@ tools       : list()</span></span>
<span></span>
<span><span class="co"># Simply look at the data after normalization</span></span>
<span><span class="co"># par(mfrow = c(1,2))</span></span>
<span><span class="co"># hist(colSums(pbmc$RNA@counts@i),breaks = 50)</span></span>
<span><span class="co"># hist(colSums(pbmc$RNA@data@i),breaks = 50)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="highly-variable-features" class="level3"><h3 class="anchored" data-anchor-id="highly-variable-features">Highly variable features</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Identify the 10 most highly variable genes</span></span>
<span><span class="va">top10</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot variable features with and without labels</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VariableFeaturePlot.html">VariableFeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LabelPoints.html">LabelPoints</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">plot1</span>, points <span class="op">=</span> <span class="va">top10</span>, repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="scaling-the-data" class="level3"><h3 class="anchored" data-anchor-id="scaling-the-data">Scaling the data</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">all_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">all_genes</span><span class="op">)</span></span>
<span><span class="co"># Remove unwanted sources of variation</span></span>
<span><span class="co"># pbmc &lt;- ScaleData(pbmc, vars.to.regress = "percent.mt")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="perform-linear-dimensional-reduction" class="level2"><h2 class="anchored" data-anchor-id="perform-linear-dimensional-reduction">Perform linear dimensional reduction</h2>
<section id="pca" class="level3"><h3 class="anchored" data-anchor-id="pca">PCA</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Examine and visualize PCA results a few different ways</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"pca"</span><span class="op">]</span><span class="op">]</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, nfeatures <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">## PC_ 1 </span></span>
<span><span class="co">## Positive:  CST3, TYROBP, LST1, AIF1, FTL </span></span>
<span><span class="co">## Negative:  MALAT1, LTB, IL32, IL7R, CD2 </span></span>
<span><span class="co">## PC_ 2 </span></span>
<span><span class="co">## Positive:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DQB1 </span></span>
<span><span class="co">## Negative:  NKG7, PRF1, CST7, GZMB, GZMA </span></span>
<span><span class="co">## PC_ 3 </span></span>
<span><span class="co">## Positive:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, HLA-DPB1 </span></span>
<span><span class="co">## Negative:  PPBP, PF4, SDPR, SPARC, GNG11 </span></span>
<span><span class="co">## PC_ 4 </span></span>
<span><span class="co">## Positive:  HLA-DQA1, CD79B, CD79A, MS4A1, HLA-DQB1 </span></span>
<span><span class="co">## Negative:  VIM, IL7R, S100A6, IL32, S100A8 </span></span>
<span><span class="co">## PC_ 5 </span></span>
<span><span class="co">## Positive:  GZMB, NKG7, S100A8, FGFBP2, GNLY </span></span>
<span><span class="co">## Negative:  LTB, IL7R, CKB, VIM, MS4A7</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="visualize-it" class="level3"><h3 class="anchored" data-anchor-id="visualize-it">Visualize it</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VizDimLoadings.html">VizDimLoadings</a></span><span class="op">(</span></span>
<span>    <span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, </span>
<span>    nfeatures <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    reduction <span class="op">=</span> <span class="st">"pca"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># PCA dotplot</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co"># PCA heatmap</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span>, cells <span class="op">=</span> <span class="fl">500</span>, balanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-7-3.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, cells <span class="op">=</span> <span class="fl">500</span>, balanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-7-4.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="determine-the-dimensionality-of-the-dataset" class="level2"><h2 class="anchored" data-anchor-id="determine-the-dimensionality-of-the-dataset">Determine the dimensionality of the dataset</h2>
<section id="jackstrawplot" class="level3"><h3 class="anchored" data-anchor-id="jackstrawplot">JackStrawPlot</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># NOTE: This process can take a long time for big datasets, comment out for expediency. More</span></span>
<span><span class="co"># approximate techniques such as those implemented in ElbowPlot() can be used to reduce</span></span>
<span></span>
<span><span class="co"># Computation time</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/JackStraw.html">JackStraw</a></span><span class="op">(</span><span class="va">pbmc</span>, num.replicate <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScoreJackStraw.html">ScoreJackStraw</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/JackStrawPlot.html">JackStrawPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="elbow-plot" class="level3"><h3 class="anchored" data-anchor-id="elbow-plot">Elbow plot</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="significant-related-genes" class="level3"><h3 class="anchored" data-anchor-id="significant-related-genes">Significant related genes</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Returns a set of genes, based on the JackStraw analysis, that have statistically significant associations with a set of PCs.</span></span>
<span><span class="co"># ?PCASigGenes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/PCASigGenes.html">PCASigGenes</a></span><span class="op">(</span><span class="va">pbmc</span>,pcs.use<span class="op">=</span><span class="fl">2</span>,pval.cut <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## [1] "PPBP"   "LYZ"    "S100A9" "IGLL5"  "GNLY"   "FTL"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="cluster-the-cells" class="level2"><h2 class="anchored" data-anchor-id="cluster-the-cells">Cluster the cells</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">pbmc</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 2638</span></span>
<span><span class="co">## Number of edges: 95927</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.8728</span></span>
<span><span class="co">## Number of communities: 9</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span>
<span></span>
<span><span class="co"># Look at cluster IDs of the first 5 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">## AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 </span></span>
<span><span class="co">##                2                3                2                1 </span></span>
<span><span class="co">## AAACCGTGTATGCG-1 </span></span>
<span><span class="co">##                6 </span></span>
<span><span class="co">## Levels: 0 1 2 3 4 5 6 7 8</span></span>
<span></span>
<span><span class="co"># Look at the cells of specific cluster</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">active.ident</span><span class="op">)</span>,<span class="va">pbmc</span><span class="op">@</span><span class="va">active.ident</span><span class="op">==</span><span class="st">"2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##                  pbmc@active.ident</span></span>
<span><span class="co">## AAACATACAACCAC-1                 2</span></span>
<span><span class="co">## AAACATTGATCAGC-1                 2</span></span>
<span><span class="co">## AAACGCACTGGTAC-1                 2</span></span>
<span><span class="co">## AAAGAGACGAGATA-1                 2</span></span>
<span><span class="co">## AAAGCCTGTATGCG-1                 2</span></span>
<span><span class="co">## AAATCAACTCGCAA-1                 2</span></span>
<span></span>
<span><span class="co"># Retrieve the cells of a cluster</span></span>
<span><span class="va">subpbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pbmc</span>,idents<span class="op">=</span><span class="st">"2"</span><span class="op">)</span></span>
<span><span class="va">subpbmc</span></span>
<span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 13714 features across 476 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (13714 features, 2000 variable features)</span></span>
<span><span class="co">##  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">##  1 dimensional reduction calculated: pca</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">subpbmc</span><span class="op">@</span><span class="va">active.ident</span>,<span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">## AAACATACAACCAC-1 AAACATTGATCAGC-1 AAACGCACTGGTAC-1 AAAGAGACGAGATA-1 </span></span>
<span><span class="co">##                2                2                2                2 </span></span>
<span><span class="co">## AAAGCCTGTATGCG-1 </span></span>
<span><span class="co">##                2 </span></span>
<span><span class="co">## Levels: 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="run-non-linear-dimensional-reduction" class="level2"><h2 class="anchored" data-anchor-id="run-non-linear-dimensional-reduction">Run non-linear dimensional reduction</h2>
<section id="umap" class="level3"><h3 class="anchored" data-anchor-id="umap">UMAP</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># If you haven't installed UMAP, you can do so via </span></span>
<span><span class="co"># reticulate::py_install(packages = "umap-learn")</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Note that you can set `label = TRUE` or use the LabelClusters function to help label</span></span>
<span><span class="co"># Individual clusters</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="tsne" class="level3"><h3 class="anchored" data-anchor-id="tsne">tSNE</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunTSNE.html">RunTSNE</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">tsne</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">)</span></span>
<span><span class="co">##                      tSNE_1     tSNE_2</span></span>
<span><span class="co">## AAACATACAACCAC-1 -12.721811   6.420117</span></span>
<span><span class="co">## AAACATTGAGCTAC-1 -20.682526 -22.307703</span></span>
<span><span class="co">## AAACATTGATCAGC-1  -3.067779  23.686369</span></span>
<span><span class="co">## AAACCGTGCTTCCG-1  30.350720  -9.899162</span></span>
<span><span class="co">## AAACCGTGTATGCG-1 -35.994115   9.507508</span></span>
<span><span class="co">## AAACGCACTGGTAC-1  -3.124182  12.680105</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section><section id="compare" class="level3"><h3 class="anchored" data-anchor-id="compare">Compare</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Note that you can set `label = TRUE` or use the LabelClusters function to help label</span></span>
<span><span class="co"># Individual clusters</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="finding-cluster-biomarkers" class="level2"><h2 class="anchored" data-anchor-id="finding-cluster-biomarkers">Finding cluster biomarkers</h2>
<section id="find-clusters" class="level3"><h3 class="anchored" data-anchor-id="find-clusters">Find clusters</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Find all markers of cluster 2</span></span>
<span><span class="va">cluster2_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster2_markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##             p_val avg_log2FC pct.1 pct.2    p_val_adj</span></span>
<span><span class="co">## IL32 2.892340e-90  1.3070772 0.947 0.465 3.966555e-86</span></span>
<span><span class="co">## LTB  1.060121e-86  1.3312674 0.981 0.643 1.453850e-82</span></span>
<span><span class="co">## CD3D 8.794641e-71  1.0597620 0.922 0.432 1.206097e-66</span></span>
<span><span class="co">## IL7R 3.516098e-68  1.4377848 0.750 0.326 4.821977e-64</span></span>
<span><span class="co">## LDHB 1.642480e-67  0.9911924 0.954 0.614 2.252497e-63</span></span>
<span></span>
<span><span class="co"># Find all markers distinguishing cluster 5 from clusters 0 and 3</span></span>
<span><span class="va">cluster5_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">5</span>, ident.2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster5_markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##                       p_val avg_log2FC pct.1 pct.2     p_val_adj</span></span>
<span><span class="co">## FCGR3A        8.246578e-205   6.794969 0.975 0.040 1.130936e-200</span></span>
<span><span class="co">## IFITM3        1.677613e-195   6.192558 0.975 0.049 2.300678e-191</span></span>
<span><span class="co">## CFD           2.401156e-193   6.015172 0.938 0.038 3.292945e-189</span></span>
<span><span class="co">## CD68          2.900384e-191   5.530330 0.926 0.035 3.977587e-187</span></span>
<span><span class="co">## RP11-290F20.3 2.513244e-186   6.297999 0.840 0.017 3.446663e-182</span></span>
<span></span>
<span><span class="co"># Find markers for every cluster compared to all remaining cells, report only the positive ones</span></span>
<span><span class="va">pbmc_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pbmc_markers</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/filter.html">filter</a></span><span class="op">(</span><span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">## # A tibble: 7,019 Ã— 7</span></span>
<span><span class="co">## # Groups:   cluster [9]</span></span>
<span><span class="co">##        p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene     </span></span>
<span><span class="co">##        &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;    </span></span>
<span><span class="co">##  1 3.75e-112       1.21 0.912 0.592 5.14e-108 0       LDHB     </span></span>
<span><span class="co">##  2 9.57e- 88       2.40 0.447 0.108 1.31e- 83 0       CCR7     </span></span>
<span><span class="co">##  3 1.15e- 76       1.06 0.845 0.406 1.58e- 72 0       CD3D     </span></span>
<span><span class="co">##  4 1.12e- 54       1.04 0.731 0.4   1.54e- 50 0       CD3E     </span></span>
<span><span class="co">##  5 1.35e- 51       2.14 0.342 0.103 1.86e- 47 0       LEF1     </span></span>
<span><span class="co">##  6 1.94e- 47       1.20 0.629 0.359 2.66e- 43 0       NOSIP    </span></span>
<span><span class="co">##  7 2.81e- 44       1.53 0.443 0.185 3.85e- 40 0       PIK3IP1  </span></span>
<span><span class="co">##  8 6.27e- 43       1.99 0.33  0.112 8.60e- 39 0       PRKCQ-AS1</span></span>
<span><span class="co">##  9 1.16e- 40       2.70 0.2   0.04  1.59e- 36 0       FHIT     </span></span>
<span><span class="co">## 10 1.34e- 34       1.96 0.268 0.087 1.84e- 30 0       MAL      </span></span>
<span><span class="co">## # â„¹ 7,009 more rows</span></span>
<span><span class="co"># ?FindAllMarkers</span></span>
<span></span>
<span><span class="va">cluster0_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span></span>
<span>    <span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">0</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>, test.use <span class="op">=</span> <span class="st">"roc"</span>, </span>
<span>    only.pos <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="visualization" class="level3"><h3 class="anchored" data-anchor-id="visualization">Visualization</h3>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">### Show expression probability distributions across clusters</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># You can plot raw counts as well</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NKG7"</span>, <span class="st">"PF4"</span><span class="op">)</span>, layer <span class="op">=</span> <span class="st">"counts"</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co"># Visualizes feature expression on a tSNE or PCA plot</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, </span>
<span>        <span class="st">"LYZ"</span>, <span class="st">"PPBP"</span>,<span class="st">"CD8A"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-3.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span></span>
<span><span class="co"># Expression heatmap for given cells and features</span></span>
<span><span class="va">top10</span> <span class="op">&lt;-</span> <span class="va">pbmc_markers</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/top_n.html">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">top10</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-16-4.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="assign-cell-type-identity-to-clusters" class="level2"><h2 class="anchored" data-anchor-id="assign-cell-type-identity-to-clusters">Assign cell type identity to clusters</h2>
<p>Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:</p>
<p>Cluster ID Markers Cell Type 0 IL7R, CCR7 Naive CD4+ T 1 CD14, LYZ CD14+ Mono 2 IL7R, S100A4 Memory CD4+ 3 MS4A1 B 4 CD8A CD8+ T 5 FCGR3A, MS4A7 FCGR3A+ Mono 6 GNLY, NKG7 NK 7 FCER1A, CST3 DC 8 PPBP Platelet</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">new_cluster_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"Naive CD4 T"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"Memory CD4 T"</span>, <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, <span class="st">"FCGR3A+ Mono"</span>,<span class="st">"NK"</span>, <span class="st">"DC"</span>, <span class="st">"Platelet"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">new_cluster_ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">RenameIdents</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">new_cluster_ids</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="index_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># p &lt;- DimPlot(pbmc, reduction = "umap", label = TRUE, label.size = 4.5) + </span></span>
<span><span class="co">#     xlab("UMAP 1") + ylab("UMAP 2") +</span></span>
<span><span class="co">#     theme(</span></span>
<span><span class="co">#         axis.title = element_text(size = 18), </span></span>
<span><span class="co">#         legend.text = element_text(size = 18)) + </span></span>
<span><span class="co">#         guides(colour = guide_legend(override.aes = list(size = 10))</span></span>
<span><span class="co">#     )</span></span>
<span><span class="co"># Save plot</span></span>
<span><span class="co"># ggsave(</span></span>
<span><span class="co">#     filename = "../output/images/pbmc3k_umap.jpg", </span></span>
<span><span class="co">#     height = 7, width = 12, plot = p, quality = 50</span></span>
<span><span class="co"># )</span></span>
<span><span class="co"># Save data</span></span>
<span><span class="co"># saveRDS(pbmc, file = "./learn/pbmc3k/pbmc3k_final.rds")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Asia/Singapore</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] patchwork_1.1.3             ggrepel_0.9.4              </span></span>
<span><span class="co">##  [3] SingleR_2.2.0               SummarizedExperiment_1.30.2</span></span>
<span><span class="co">##  [5] Biobase_2.62.0              GenomicRanges_1.52.1       </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.38.0         IRanges_2.36.0             </span></span>
<span><span class="co">##  [9] S4Vectors_0.40.2            BiocGenerics_0.48.1        </span></span>
<span><span class="co">## [11] MatrixGenerics_1.14.0       matrixStats_1.1.0          </span></span>
<span><span class="co">## [13] Seurat_5.0.1                SeuratObject_5.0.1         </span></span>
<span><span class="co">## [15] sp_2.1-2                    lubridate_1.9.3            </span></span>
<span><span class="co">## [17] forcats_1.0.0               stringr_1.5.1              </span></span>
<span><span class="co">## [19] dplyr_1.1.4                 purrr_1.0.2                </span></span>
<span><span class="co">## [21] readr_2.1.4                 tidyr_1.3.0                </span></span>
<span><span class="co">## [23] tibble_3.2.1                ggplot2_3.4.4              </span></span>
<span><span class="co">## [25] tidyverse_2.0.0             here_1.0.1                 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RcppAnnoy_0.0.21          splines_4.3.1            </span></span>
<span><span class="co">##   [3] later_1.3.1               bitops_1.0-7             </span></span>
<span><span class="co">##   [5] R.oo_1.25.0               polyclip_1.10-6          </span></span>
<span><span class="co">##   [7] fastDummies_1.7.3         lifecycle_1.0.4          </span></span>
<span><span class="co">##   [9] rprojroot_2.0.4           globals_0.16.2           </span></span>
<span><span class="co">##  [11] lattice_0.22-5            MASS_7.3-60              </span></span>
<span><span class="co">##  [13] magrittr_2.0.3            limma_3.58.1             </span></span>
<span><span class="co">##  [15] plotly_4.10.3             rmarkdown_2.25           </span></span>
<span><span class="co">##  [17] yaml_2.3.7                httpuv_1.6.12            </span></span>
<span><span class="co">##  [19] sctransform_0.4.1         spam_2.10-0              </span></span>
<span><span class="co">##  [21] spatstat.sparse_3.0-3     reticulate_1.34.0        </span></span>
<span><span class="co">##  [23] cowplot_1.1.1             pbapply_1.7-2            </span></span>
<span><span class="co">##  [25] RColorBrewer_1.1-3        abind_1.4-5              </span></span>
<span><span class="co">##  [27] zlibbioc_1.48.0           Rtsne_0.16               </span></span>
<span><span class="co">##  [29] R.utils_2.12.3            RCurl_1.98-1.13          </span></span>
<span><span class="co">##  [31] GenomeInfoDbData_1.2.11   irlba_2.3.5.1            </span></span>
<span><span class="co">##  [33] listenv_0.9.0             spatstat.utils_3.0-4     </span></span>
<span><span class="co">##  [35] goftest_1.2-3             RSpectra_0.16-1          </span></span>
<span><span class="co">##  [37] spatstat.random_3.2-2     fitdistrplus_1.1-11      </span></span>
<span><span class="co">##  [39] parallelly_1.36.0         DelayedMatrixStats_1.22.6</span></span>
<span><span class="co">##  [41] leiden_0.4.3.1            codetools_0.2-19         </span></span>
<span><span class="co">##  [43] DelayedArray_0.26.7       tidyselect_1.2.0         </span></span>
<span><span class="co">##  [45] farver_2.1.1              ScaledMatrix_1.8.1       </span></span>
<span><span class="co">##  [47] spatstat.explore_3.2-5    jsonlite_1.8.7           </span></span>
<span><span class="co">##  [49] ellipsis_0.3.2            progressr_0.14.0         </span></span>
<span><span class="co">##  [51] ggridges_0.5.4            survival_3.5-7           </span></span>
<span><span class="co">##  [53] tools_4.3.1               ica_1.0-3                </span></span>
<span><span class="co">##  [55] Rcpp_1.0.11               glue_1.6.2               </span></span>
<span><span class="co">##  [57] gridExtra_2.3             xfun_0.41                </span></span>
<span><span class="co">##  [59] withr_2.5.2               fastmap_1.1.1            </span></span>
<span><span class="co">##  [61] fansi_1.0.5               digest_0.6.33            </span></span>
<span><span class="co">##  [63] rsvd_1.0.5                timechange_0.2.0         </span></span>
<span><span class="co">##  [65] R6_2.5.1                  mime_0.12                </span></span>
<span><span class="co">##  [67] colorspace_2.1-0          scattermore_1.2          </span></span>
<span><span class="co">##  [69] tensor_1.5                spatstat.data_3.0-3      </span></span>
<span><span class="co">##  [71] R.methodsS3_1.8.2         utf8_1.2.4               </span></span>
<span><span class="co">##  [73] generics_0.1.3            data.table_1.14.8        </span></span>
<span><span class="co">##  [75] httr_1.4.7                htmlwidgets_1.6.3        </span></span>
<span><span class="co">##  [77] S4Arrays_1.0.6            uwot_0.1.16              </span></span>
<span><span class="co">##  [79] pkgconfig_2.0.3           gtable_0.3.4             </span></span>
<span><span class="co">##  [81] lmtest_0.9-40             XVector_0.42.0           </span></span>
<span><span class="co">##  [83] htmltools_0.5.7           dotCall64_1.1-1          </span></span>
<span><span class="co">##  [85] scales_1.3.0              png_0.1-8                </span></span>
<span><span class="co">##  [87] knitr_1.45                tzdb_0.4.0               </span></span>
<span><span class="co">##  [89] reshape2_1.4.4            nlme_3.1-163             </span></span>
<span><span class="co">##  [91] zoo_1.8-12                KernSmooth_2.23-22       </span></span>
<span><span class="co">##  [93] parallel_4.3.1            miniUI_0.1.1.1           </span></span>
<span><span class="co">##  [95] vipor_0.4.5               ggrastr_1.0.2            </span></span>
<span><span class="co">##  [97] pillar_1.9.0              grid_4.3.1               </span></span>
<span><span class="co">##  [99] vctrs_0.6.5               RANN_2.6.1               </span></span>
<span><span class="co">## [101] promises_1.2.1            BiocSingular_1.16.0      </span></span>
<span><span class="co">## [103] beachmat_2.16.0           xtable_1.8-4             </span></span>
<span><span class="co">## [105] cluster_2.1.4             beeswarm_0.4.0           </span></span>
<span><span class="co">## [107] evaluate_0.23             cli_3.6.1                </span></span>
<span><span class="co">## [109] compiler_4.3.1            rlang_1.1.2              </span></span>
<span><span class="co">## [111] crayon_1.5.2              future.apply_1.11.0      </span></span>
<span><span class="co">## [113] labeling_0.4.3            plyr_1.8.9               </span></span>
<span><span class="co">## [115] ggbeeswarm_0.7.2          stringi_1.8.2            </span></span>
<span><span class="co">## [117] viridisLite_0.4.2         deldir_2.0-2             </span></span>
<span><span class="co">## [119] BiocParallel_1.36.0       munsell_0.5.0            </span></span>
<span><span class="co">## [121] lazyeval_0.2.2            spatstat.geom_3.2-7      </span></span>
<span><span class="co">## [123] Matrix_1.6-3              RcppHNSW_0.5.0           </span></span>
<span><span class="co">## [125] hms_1.1.3                 sparseMatrixStats_1.12.2 </span></span>
<span><span class="co">## [127] future_1.33.0             statmod_1.5.0            </span></span>
<span><span class="co">## [129] shiny_1.8.0               ROCR_1.0-11              </span></span>
<span><span class="co">## [131] igraph_1.5.1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/guoruizhong\.github\.io\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p><span class="faux-block"><i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> 2021-2024 Guorui Zhong ~ Made with <a href="https://www.r-project.org/"><i class="fa-brands fa-r-project" aria-label="r-project"></i></a> and <a href="https://quarto.org/">Quarto</a></span></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>


</body></html>